### ISSUE TITLE
GNU Radio Hardware Integration for Kaira

### ISSUE THREAD

**Issue by selimfirat** [ipc-lab/kaira#24]
# GNU Radio Hardware Integration for Kaira

## Issue Type
- [x] Feature Request
- [ ] Bug Report
- [ ] Enhancement
- [ ] Documentation

## Summary
Implement GNU Radio bindings to enable hardware-in-the-loop testing and real-world deployment of Kaira's PyTorch-based communication models on Software Defined Radio (SDR) platforms.

## Motivation
Kaira currently provides excellent simulation capabilities for deep learning-based communication systems (particularly DeepJSCC), but lacks integration with real hardware for validation and deployment. Adding GNU Radio bindings would:

1. **Bridge the simulation-to-reality gap** by enabling direct hardware testing of trained models
2. **Validate model performance** in real wireless environments with actual channel impairments
3. **Enable practical deployment** of deep learning communication systems
4. **Support research reproducibility** with standardized hardware interfaces
5. **Facilitate rapid prototyping** of novel communication algorithms

#### 1. Hardware Module Structure
```
kaira/hardware/
├── __init__.py              # Module exports
├── sdr_utils.py            # SDR configuration and utilities
├── gnuradio_bridge.py      # Main GNU Radio integration
├── examples.py             # Usage examples
└── README.md               # Hardware-specific documentation
```

#### 2. Key Classes and Interfaces

**SDRConfig**: Configuration management for SDR parameters
```python
@dataclass
class SDRConfig:
    center_frequency: float     # Operating frequency (Hz)
    sample_rate: float         # Sample rate (Hz) 
    tx_gain: float = 0.0       # Transmit gain (dB)
    rx_gain: float = 0.0       # Receive gain (dB)
    bandwidth: Optional[float] = None
    device_args: Optional[str] = None
    # ... additional parameters
```

**GNURadioBridge**: Main integration class
```python
class GNURadioBridge:
    def create_flowgraph(self) -> gr.top_block
    def add_torch_encoder(self, model: torch.nn.Module) -> TorchEncoderBlock
    def add_torch_decoder(self, model: torch.nn.Module) -> TorchDecoderBlock
    def add_source(self, source_type: str, **kwargs) -> Any
    def add_sink(self, sink_type: str, **kwargs) -> Any
    def connect(self, src, dst, src_port: int = 0, dst_port: int = 0)
    def start(self) / stop(self)
```

**TorchEncoderBlock/TorchDecoderBlock**: GNU Radio blocks wrapping PyTorch models
```python
class TorchEncoderBlock(gr.sync_block):
    def work(self, input_items, output_items):
        # Convert GNU Radio data → PyTorch tensors → process → convert back
```

#### 3. Integration Points

- **Model Compatibility**: Support for existing Kaira encoder/decoder models
- **Real-time Processing**: Efficient tensor conversion and GPU utilization
- **Hardware Abstraction**: Support for USRP, RTL-SDR, BladeRF, HackRF
- **Channel Integration**: Interface with Kaira's channel models for comparison

#### Dependencies
- **GNU Radio ≥ 3.9.0**: Core SDR framework
- **UHD**: USRP hardware drivers
- **PyTorch**: Existing Kaira dependency
- **NumPy/SciPy**: Signal processing utilities

#### Hardware Support
- **Primary**: USRP B200/B210, N200/N210 series
- **Secondary**: RTL-SDR (RX only), BladeRF, HackRF
- **Frequency Range**: 70 MHz - 6 GHz (hardware dependent)

#### Performance Targets
- **Real-time processing** at sample rates up to 10 MHz
- **GPU acceleration** support for inference
- **Low latency** data conversion (<1ms for typical buffer sizes)
- **Memory efficiency** for continuous operation

### ✅ Completed
- [x] Core module structure and API design
- [x] SDR configuration management (`SDRConfig`, `FrequencyBand`)
- [x] Basic GNU Radio bridge implementation
- [x] PyTorch encoder/decoder block wrappers
- [x] Utility functions (frequency validation, dB conversions)
- [x] Example usage scripts
- [x] Initial test suite
- [x] Documentation framework

### 🚧 In Progress  
- [ ] **Error handling robustness** - Graceful degradation when GNU Radio unavailable
- [ ] **Real-time optimization** - Buffer management and threading
- [ ] **Hardware detection** - Automatic SDR device discovery

### 📋 Pending
- [ ] **Hardware validation** - Testing with actual SDR devices
- [ ] **Performance benchmarking** - Latency and throughput measurements  
- [ ] **Advanced examples** - Complete end-to-end scenarios
- [ ] **Integration tests** - Automated testing with simulated hardware
- [ ] **Documentation completion** - API reference and tutorials
- [ ] **CI/CD integration** - Automated testing pipeline

### Basic Transmission
```python
from kaira.hardware import GNURadioBridge, SDRConfig, FrequencyBand

# Configure for 915 MHz ISM band
config = SDRConfig(
    center_frequency=FrequencyBand.ISM_915.value,
    sample_rate=1e6,
    tx_gain=20.0
)

# Create bridge and flowgraph
bridge = GNURadioBridge(config)
bridge.create_flowgraph()

# Add PyTorch encoder
encoder = MyDeepJSCCEncoder()  # Your trained model
torch_encoder = bridge.add_torch_encoder(encoder)

# Connect to USRP
source = bridge.add_source("file", filename="data.bin")
sink = bridge.add_sink("usrp")

bridge.connect(source, torch_encoder)
bridge.connect(torch_encoder, sink)
bridge.start()
```

### Hardware-in-the-Loop Testing
```python
# Transmitter side
tx_bridge = GNURadioTransmitter(config, encoder_model)
tx_bridge.start()

# Receiver side  
rx_bridge = GNURadioReceiver(config, decoder_model)
rx_bridge.start()

# Validate end-to-end performance
evaluate_ber_performance(tx_bridge, rx_bridge)
```

### Unit Tests
- [x] Configuration validation
- [x] Utility function correctness
- [x] Mock GNU Radio integration
- [ ] Error handling scenarios

### Integration Tests  
- [ ] File-based signal processing (no hardware required)
- [ ] Simulated hardware testing
- [ ] Real hardware validation (manual)

### Performance Tests
- [ ] Latency measurements
- [ ] Throughput benchmarking  
- [ ] Memory usage profiling
- [ ] GPU utilization analysis

### User Documentation
- [x] **API Reference** - Complete class and method documentation
- [x] **Quick Start Guide** - Basic setup and usage
- [ ] **Hardware Setup** - Device-specific configuration guides
- [ ] **Troubleshooting** - Common issues and solutions
- [ ] **Performance Tuning** - Optimization guidelines

### Developer Documentation
- [ ] **Architecture Overview** - System design and data flow
- [ ] **Contributing Guide** - Development workflow and standards  
- [ ] **Testing Guide** - How to test without hardware
- [ ] **Extension Guide** - Adding support for new hardware

### GNU Radio Versions
- **Primary**: GNU Radio 3.10+ (latest stable)
- **Secondary**: GNU Radio 3.9+ (backward compatibility)
- **Deprecation**: GNU Radio 3.8 and earlier

### Python Compatibility
- **Supported**: Python 3.8, 3.9, 3.10, 3.11
- **Testing Priority**: Python 3.9+ (matches Kaira requirements)

### Operating Systems
- **Primary**: Ubuntu 20.04/22.04 LTS
- **Secondary**: Fedora, CentOS, macOS
- **Limited**: Windows (GNU Radio installation complexity)

### Technical Risks
- **Real-time constraints**: PyTorch inference may not meet timing requirements
  - *Mitigation*: GPU acceleration, model optimization, buffer tuning
- **GNU Radio complexity**: Integration challenges with complex flowgraphs  
  - *Mitigation*: Start with simple blocks, incremental complexity
- **Hardware variability**: Different SDR devices have varying capabilities
  - *Mitigation*: Abstraction layer, device-specific profiles

### Dependency Risks
- **GNU Radio installation**: Complex dependency management
  - *Mitigation*: Optional dependency, graceful degradation, container support
- **Hardware drivers**: Platform-specific driver issues
  - *Mitigation*: Clear documentation, community support

### Functional Metrics
- [ ] **Basic operation**: Successfully transmit/receive with USRP
- [ ] **Model integration**: Deploy existing Kaira models to hardware
- [ ] **Performance validation**: Achieve <5% BER increase vs simulation

### Usability Metrics  
- [ ] **Setup time**: <30 minutes from installation to first transmission
- [ ] **Documentation quality**: Users can follow examples without support
- [ ] **Error recovery**: Clear error messages and debugging guidance

### Community Metrics
- [ ] **Adoption**: 10+ GitHub stars, 5+ issues/PRs from community
- [ ] **Integration**: Used in 3+ research papers or projects
- [ ] **Maintenance**: Monthly activity, responsive issue resolution

### Phase 1: Core Functionality (2-3 weeks)
- [ ] Robust error handling and graceful degradation
- [ ] Hardware validation with basic USRP setup
- [ ] Performance optimization for real-time operation

### Phase 2: Advanced Features (3-4 weeks)  
- [ ] Support for additional hardware (RTL-SDR, BladeRF)
- [ ] Advanced GNU Radio block implementations
- [ ] Comprehensive test suite with CI integration

### Phase 3: Documentation & Examples (2-3 weeks)
- [ ] Complete API documentation
- [ ] Hardware setup guides
- [ ] Advanced usage examples and tutorials

### Phase 4: Community & Maintenance (Ongoing)
- [ ] Community feedback integration
- [ ] Bug fixes and performance improvements
- [ ] New hardware support based on user requests

### GNU Radio Resources
- [GNU Radio Documentation](https://www.gnuradio.org/doc/)
- [GNU Radio Tutorials](https://wiki.gnuradio.org/index.php/Tutorials)
- [PyBOMBS Installation](https://github.com/gnuradio/pybombs)

### Similar Projects
- **gr-pytorch**: Early-stage PyTorch integration for GNU Radio
- **LimeSDR-Mini**: Another SDR platform with Python bindings
- **OpenBTS**: Software-defined cellular base station

### Research Applications
- Deep learning for wireless communications
- Neural network-based channel coding
- Cognitive radio and spectrum sensing
- Massive MIMO and beamforming

## Call for Collaboration

This is a significant feature that would benefit from community involvement:

### Seeking Contributors For:
- **Hardware testing** with different SDR platforms
- **Performance optimization** and real-time processing expertise  
- **Documentation** and tutorial creation
- **Example development** for specific use cases

### Hardware Donations Welcome:
- USRP devices for testing and development
- Other SDR platforms (BladeRF, HackRF, etc.)
- RF test equipment for validation

## Labels
`enhancement` `hardware` `gnu-radio` `sdr` `integration` `help-wanted` `good-first-issue` (for testing tasks)

---

**Note**: This implementation builds upon Kaira's existing modular architecture and maintains compatibility with current PyTorch-based workflows. The goal is to provide a seamless bridge between simulation and real-world deployment while preserving Kaira's ease of use and research focus.
